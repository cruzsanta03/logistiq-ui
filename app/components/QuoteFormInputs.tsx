import React from "react"

type Props = {
  form: Record<string, string | number>
  handleChange: (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void
  loadingDistance: boolean
  formatCurrency: (val: string | number) => string
}

const QuoteFormInputs: React.FC<Props> = ({ form, handleChange, loadingDistance, formatCurrency }) => {
  const [errors, setErrors] = React.useState<Record<string, string>>({});

  return (
    <div className="grid gap-4 p-6 border rounded-md shadow bg-white">
      <div className="text-sm text-gray-700">
        <p className="mb-2">
          <strong>Roundtrip Miles:</strong> {form.roundtrip || "N/A"}{" "}
          {loadingDistance && <span className="text-blue-500">(Calculating...)</span>}
        </p>
        <label htmlFor="fsc" className="block font-semibold">Fuel Surcharge %</label>
        <input
          id="fsc"
          type="number"
          name="fsc"
          value={form.fsc as string | number}
          onChange={handleChange}
          className="border px-2 py-1 rounded w-24"
        />
        <span className="ml-4">
          <strong>Fuel:</strong> {formatCurrency(form.fuel)}
        </span>
      </div>

      {[
        "company", "destination", "requestor", "linehaul", "prePull", "chassis", "stopOff", "storage",
        "hazMat", "reefer", "dryRun", "manualAdjustment", "specialInstructions", "promoCode", "validThrough"
      ].map((field) => (
        <div key={field} className="flex flex-col">
          <input
            name={field}
            placeholder={field.replace(/([A-Z])/g, " $1").replace(/^./, (str) => str.toUpperCase())}
            type={
              [
                "manualAdjustment", "prePull", "chassis", "stopOff",
                "storage", "hazMat", "reefer", "dryRun"
              ].includes(field)
                ? "number"
                : field === "validThrough"
                ? "date"
                : "text"
            }
            value={form[field] as string | number}
            onChange={handleChange}
            onBlur={(e) => {
              if (!e.target.value.trim()) {
                setErrors((prev) => ({ ...prev, [e.target.name]: "This field is required" }));
              } else {
                const { [e.target.name]: _, ...rest } = errors;
                setErrors(rest);
              }
            }}
            className="border px-2 py-1 rounded"
          />
          {errors[field] && (
            <p className="text-red-600 text-sm">{errors[field]}</p>
          )}
        </div>
      ))}
    </div>
  );
};

export default QuoteFormInputs;
