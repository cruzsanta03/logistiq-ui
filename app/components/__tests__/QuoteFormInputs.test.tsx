import React from "react"
import { render, screen, fireEvent } from "@testing-library/react"
import "@testing-library/jest-dom"
import userEvent from "@testing-library/user-event"
import renderer from "react-test-renderer"
import QuoteFormInputs from "../QuoteFormInputs"

const baseProps = {
  form: {
    fsc: 5,
    fuel: 100,
    roundtrip: 200,
    company: "",
    destination: "",
    requestor: "",
    linehaul: "",
    prePull: "",
    chassis: "",
    stopOff: "",
    storage: "",
    hazMat: "",
    reefer: "",
    dryRun: "",
    manualAdjustment: "",
    specialInstructions: "",
    promoCode: "",
    validThrough: ""
  },
  handleChange: jest.fn(),
  loadingDistance: false,
  formatCurrency: function (val: string | number) {
    return `$${val}`;
  }
};

describe("QuoteFormInputs", () => {
  it("renders Roundtrip Miles and Fuel Surcharge fields", () => {
    render(<QuoteFormInputs {...baseProps} />);
    expect(screen.getByText(/Roundtrip Miles:/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Fuel Surcharge %/i)).toBeInTheDocument();
  });

  it("calls handleChange when FSC input changes", () => {
    render(<QuoteFormInputs {...baseProps} />);
    const input = screen.getByLabelText(/Fuel Surcharge %/i);
    fireEvent.change(input, { target: { value: "10" } });
    expect(baseProps.handleChange).toHaveBeenCalled();
  });

  it("renders dynamic fields from map()", () => {
    render(<QuoteFormInputs {...baseProps} />);
    expect(screen.getByPlaceholderText("Company")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Promo Code")).toBeInTheDocument();
  });

  it("formats currency correctly", () => {
    const result = baseProps.formatCurrency(123);
    expect(result).toBe("$123");
  });

  it("shows 'Calculating...' when loadingDistance is true", () => {
    render(<QuoteFormInputs {...baseProps} loadingDistance={true} />);
    expect(screen.getByText("(Calculating...)")).toBeInTheDocument();
  });

  it("renders correct roundtrip value", () => {
    render(<QuoteFormInputs {...baseProps} />);
    expect(screen.getByText(/200/)).toBeInTheDocument();
  });

  it("renders N/A when roundtrip is missing", () => {
    const props = { ...baseProps, form: { ...baseProps.form, roundtrip: "" } };
    render(<QuoteFormInputs {...props} />);
    expect(screen.getByText(/N\/A/)).toBeInTheDocument();
  });

  it("matches the snapshot", () => {
    const tree = renderer.create(<QuoteFormInputs {...baseProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("lets the user type in company field (with state)", async () => {
    const Wrapper = () => {
      const [form, setForm] = React.useState({ ...baseProps.form });
      return (
        <QuoteFormInputs
          form={form}
          handleChange={(e) =>
            setForm({ ...form, [e.target.name]: e.target.value })
          }
          loadingDistance={false}
          formatCurrency={(val) => `$${val}`}
        />
      );
    };
    render(<Wrapper />);
    const input = screen.getByPlaceholderText("Company");
    await userEvent.type(input, "CyberCorp", { delay: 0 });
    expect(input).toHaveValue("CyberCorp");
  });
});

it("shows error on blur if required field is empty", async () => {
  const Wrapper = () => {
    const [form, setForm] = React.useState({ ...baseProps.form });
    const [errors, setErrors] = React.useState({});
    return (
      <QuoteFormInputs
        form={form}
        handleChange={(e) =>
          setForm({ ...form, [e.target.name]: e.target.value })
        }
        loadingDistance={false}
        formatCurrency={(val) => `$${val}`}
      />
    );
  };

  render(<Wrapper />);
  const input = screen.getByPlaceholderText("Company");
  await userEvent.clear(input);
  input.blur();
  expect(await screen.findByText("This field is required")).toBeInTheDocument();
});
