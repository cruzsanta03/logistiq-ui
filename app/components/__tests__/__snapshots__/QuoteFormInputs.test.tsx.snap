// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteFormInputs matches the snapshot 1`] = `null`;
